ios/MaasCore/Sources/MaasComponents/Supporting Files/CircularSpinner.swift

```diff
@@ -0,0 +1,88 @@
+struct Spinner: View {
+    
+    private static let initialDegree = Angle.degrees(135)
+    private static let initialSpinnerLength: CGFloat = 0.75
+
+    let color: Color
+    var lineWidth: CGFloat = 1.5
+    var animationTime: Double = 1
+    var appearDuration: Double = 0.2
+    @Binding var isAnimating: Bool
+    
+    @State private var spinnerStart: CGFloat = 0
+    @State private var spinnerEnd: CGFloat = initialSpinnerLength
+    @State private var rotationDegree = initialDegree
+    @State private var hasAppeared = false
+    
+    private var shouldAnimate: Bool { hasAppeared && isAnimating }
+
+    var body: some View {
+        Circle()
+            .trim(
+                from: shouldAnimate ? spinnerStart + 0.1 : spinnerStart,
+                to: shouldAnimate ? spinnerEnd - 0.2 : spinnerEnd
+            )
+            .stroke(style: StrokeStyle(lineWidth: lineWidth, lineCap: .round))
+            .fill(color)
+            .animation(lineLengthAnimation)
+            
+            .rotationEffect(shouldAnimate
+                                ? Self.initialDegree + Angle.degrees(360)
+                                : Self.initialDegree + Angle.degrees(0))
+            .animation(rotationAnimation)
+            
+            .scaleEffect(shouldAnimate ? 1 : 0)
+            .animation(.easeOut(duration: appearDuration))
+            // -
+            .padding(lineWidth / 2)
+            .onAppear { hasAppeared = true }
+            .onDisappear { hasAppeared = false }
+            .animation(nil)
+
+    }
+    
+    private var lineLengthAnimation: Animation {
+        Animation
+            .easeInOut(duration: animationTime / 2)
+            .repeat(while: shouldAnimate, autoreverses: true)
+    }
+    
+    private var rotationAnimation: Animation {
+        shouldAnimate
+            ? Animation.linear(duration: animationTime)
+            .repeat(while: shouldAnimate, autoreverses: false)
+            : Animation.default
+    }
+    
+}
+
+extension Animation {
+    
+    func `repeat`(while expression: Bool, autoreverses: Bool = false) -> Animation {
+        if expression {
+            return self.repeatForever(autoreverses: autoreverses)
+        } else {
+            return self
+        }
+    }
+}
```

Comment:
I want to say that we should move this in to separate file and make it public... But i won't say it... 🙈

Reply:
🙈